src/core/smallvec_type.hpp

```diff
@@ -7,14 +7,131 @@
+}
+
+/** Get a non-const iterator to a matching element in the vector.
+ * @param vec the vector to be searched
+ * @param elem the element to search for
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+typename std::vector<T>::iterator Find(std::vector<T>& vec, const T& elem) {
+	return std::find(vec.begin(), vec.end(), elem);
+}
+
+/** Get the index of the first matching element.
+ * @param vec the vector to be searched
+ * @param elem the element to search for
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+int FindIndex(const std::vector<T>& vec, const T& elem) {
+	auto it = Find(vec, elem);
+	if (vec.end() == it) return -1;
+	return std::distance(vec.begin(), it);
+}
+
+/** Check whether an element exists.
+ * @param vec the vector to be searched
+ * @param elem the element to search for
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+bool Contains(const std::vector<T>& vec, const T& elem) {
+	return Find(vec, elem) != vec.end();
+}
+
+/** Count instances of an element.
+ * @param vec the vector to be searched
+ * @param elem the element to count instances of
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+size_t Count(const std::vector<T>& vec, const T& elem) {
+	return std::count(vec.begin(), vec.end(), elem);
+}
+
+/** Append an element if it does not already exist in the vector.
+ * @param vec the vector to be modified
+ * @param elem the element to add for and remove if missing
+ * @tparam T the contained type of the vector
+ * @tparam U the universal reference type for perfect forwarding
+ */
+template <typename T, typename U>
+bool Include(std::vector<T>& vec, U&& elem) {
+	auto missing = not Contains(vec, elem);
+	if (missing) vec.push_back(std::forward<U>(elem));
+	return missing;
+}
+
+/** Remove an element referenced in the vector and move the last element to take its place.
+ * @param vec the vector to be modified
+ * @param it the element to remove
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+typename std::vector<T>::iterator Erase(std::vector<T>& vec, typename std::vector<T>::iterator it) {
+	assert(vec.begin() <= it and vec.end() > it);
+	if (std::prev(vec.end()) != it) std::swap(vec.back(), *it);
+	vec.pop_back();
+	return it;
+}
+
+/** Remove an element if it exists in the vector and move the last element to take its place.
+ * @param vec the vector to be modified
+ * @param elem the element to search for and remove if found
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+bool Exclude(std::vector<T>& vec, const T& elem) {
+	auto it = Find(vec, elem);
+	if (vec.end() == it) return false;
+	Erase(vec, it);
+	return true;
+}
+
+/** Clear a vector and then force it to reduce its capacity.
+ * @param vec the vector to be modified
+ * @tparam T the contained type of the vector
+ */
+template <typename T>
+void Reset(std::vector<T>& vec) {
```

Comment:
pretty sure this function isn't necessary - can be replaced with a clear() - the capacity reduction was just due to how the allocations worked

Reply:
It may be a good idea to keep the capacity reduction where it is used in case that behavior is desired.
According to [cppreference](http://en.cppreference.com/w/cpp/container/vector/clear): "Leaves the capacity() of the vector unchanged"

Reply:
`shrink_to_fit` is only a request which may be ignored.
The guaranteed implementation looks like this:
```cpp
template <typename T>
void Reset(std::vector<T>& vec) {
  std::vector<T>{}.swap(vec);
}
```